src/Akka.Streams.Kafka/Stages/Consumers/Concrete/CommittableSubSourceStage.cs

```diff
@@ -13,44 +14,85 @@
+        Source<CommittableMessage<K, V>, NotUsed>)>
     {
         private readonly Func<ConsumeResult<K, V>, string> _metadataFromRecord;
-        
+
         /// <summary>
         /// Consumer settings
         /// </summary>
         public ConsumerSettings<K, V> Settings { get; }
+
         /// <summary>
         /// Subscription
         /// </summary>
         public IAutoSubscription Subscription { get; }
 
-        public CommittableSubSourceStage(ConsumerSettings<K, V> settings, IAutoSubscription subscription, Func<ConsumeResult<K, V>, string>? metadataFromRecord = null) 
+        public CommittableSubSourceStage(ConsumerSettings<K, V> settings, IAutoSubscription subscription,
+            Func<ConsumeResult<K, V>, string>? metadataFromRecord = null)
             : base("CommittableSubSourceStage")
         {
             Settings = settings;
             Subscription = subscription;
             _metadataFromRecord = metadataFromRecord ?? (_ => string.Empty);
+            _subSourceStageLogicFactory = new CommittableSubSourceStageLogicFactory(Settings, _metadataFromRecord);
         }
 
-        protected override (GraphStageLogic, IControl) Logic(SourceShape<(TopicPartition, Source<CommittableMessage<K, V>, NotUsed>)> shape, Attributes inheritedAttributes)
+        private readonly SubSourceLogic.ISubSourceStageLogicFactory<K, V, CommittableMessage<K, V>> _subSourceStageLogicFactory;
+
+        protected override (GraphStageLogic, IControl) Logic(
+            SourceShape<(TopicPartition, Source<CommittableMessage<K, V>, NotUsed>)> shape,
+            Attributes inheritedAttributes)
         {
-            var logic = new SubSourceLogic<K, V, CommittableMessage<K, V>>(shape, Settings, Subscription, 
-                                                                           messageBuilderFactory: GetMessageBuilder, 
-                                                                           getOffsetsOnAssign: Option<Func<IImmutableSet<TopicPartition>, Task<IImmutableSet<TopicPartitionOffset>>>>.None, 
-                                                                           onRevoke: _ => { }, 
-                                                                           attributes: inheritedAttributes);
+            var logic = new SubSourceLogic<K, V, CommittableMessage<K, V>>(shape, Settings, Subscription,
+                getOffsetsOnAssign:
+                Option<Func<IImmutableSet<TopicPartition>, Task<IImmutableSet<TopicPartitionOffset>>>>.None,
+                onRevoke: _ => { },
+                _subSourceStageLogicFactory,
+                attributes: inheritedAttributes);
             return (logic, logic.Control);
         }
         
-        /// <summary>
-        /// Creates message builder for sub-source logic
-        /// </summary>
-        private CommittableSourceMessageBuilder<K, V> GetMessageBuilder(SubSourceLogic<K, V, CommittableMessage<K, V>> logic)
+        private class
```

Comment:
Implemented the `IMessageBuilder` and all commitable-source details inside the `CommittableSubSourceStageLogicFactory`